crates/ra_hir/src/ty/infer.rs

```diff
@@ -915,9 +929,18 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
             Expr::StructLit { path, fields, spread } => {
                 let (ty, def_id) = self.resolve_variant(path.as_ref());
                 let substs = ty.substs().unwrap_or_else(Substs::empty);
-                for field in fields {
+                for (field_idx, field) in fields.into_iter().enumerate() {
                     let field_ty = def_id
-                        .and_then(|it| it.field(self.db, &field.name))
+                        .and_then(|it| match it.field(self.db, &field.name) {
+                            Some(field) => Some(field),
+                            None => {
+                                self.diagnostics.push(InferenceDiagnostic::NoSuchField {
```

Comment:
we should probably only report diagnostics here if ty != Ty::Unknown

Reply:
I thought that defid being not None is a sufficient condition here.

Reply:
I'm not sure (I'm on mobile), but I think if the type is `ty::unknown` it could be because they used a construct we don't yet support (e.g. I don't think we support the `?` operator yet, as I haven't seen a PR for it).
In these cases reporting errors is bad when it's a problem with our inference.

Reply:
You're right of course, I missed that this doesn't get hit when def_id is None :)